Added missing handling for some dependency configuration keys

boesing opened the pull request:
|    Q          |   A
|-------------- | ------
| Bugfix        | yes

### Description

With laminas/laminas-zendframework-bridge#60 it came up, that we have missed the `delegators`.
After further investigation, I found out that we also dropped handling for some other keys of the `ServiceManager` configuration.

Thus, I've added the following keys to the `ConfigPostProcessor` aswell:
- handles `abstract_factories`
- handles `initializers`
- handles `lazy_services`
- handles `services`
- handles `shared`

---

After some discussions with @michalbundyra (as we both had the same issues while migrating real-world projects), we also found out that we do not handle invalid configurations.

First of all, we are just skipping that invalid configurations and keep them as they are.
In a next step, we create a new feature in `laminas/laminas-servicemanager` where one can add a `strict` flag to all `ServiceManager` configurations (including plugin managers of course), which will verify that the configuration is set correctly. In one of the next minor versions of `laminas/laminas-servicemanager`, the flag might be set to `false` per-default. In the next major, this will change to `true` per-default and throws exceptions on invalid configurations.

boesing committed:
Added failing unit tests

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
Extended output of `LazyServices`

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
Extended test case for `abstract_factories`

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
Added failing unit test for a full service manager configuration

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
Added handling for missing service manager configurations

- handles `abstract_factories`
- handles `initializers`
- handles `lazy_services`
- handles `services`
- handles `shared`

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
Added failing unit test for an invalid configuration

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
Extended invalid configuration for `invokables`

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
Added handling for invalid configurations

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing commented:
This will fix #62

weierophinney commented:
@boesing Travis is failing, and it looks like delegator lists are getting **both** the translated Laminas class references **and** the original ZF ones. Can you review and either update test expectations (and, if you do, maybe indicate _why_) to include them, or fix the underlying problem (assuming you determine what it is).

When you've done that, I'll review, merge, and release.

Thanks!

weierophinney commented:
@boesing Also, you'll need to rebase, as I merged #61.

weierophinney commented:
I've rebased and pushed back to your branch. I believe @michalbundyra has some improvements to suggest, based on feedback on #61 he pointed me to after I merged.

michalbundyra committed:
Optimisation: reuse implemented functions and replace all other data

We need apply special rules just for `aliases`, `invokables` and `factories`,
rest should be replaced the normal way.

Signed-off-by: Michał Bundyra <contact@webimpress.com>

michalbundyra committed:
CS fixes - spacing

Signed-off-by: Michał Bundyra <contact@webimpress.com>

michalbundyra committed:
Fallback to normal replacement procedure if there is no aliases, factories and invokables defined in the array


Signed-off-by: Michał Bundyra <contact@webimpress.com>

michalbundyra commented on src/ConfigPostProcessor.php:
```diff
@@ -251,9 +259,13 @@ private function replaceDependencyConfiguration(array $config)
-            $config['delegators'] = $delegators;
+
+        foreach ($config as $key => $data) {
+            if (in_array($key, ['aliases', 'invokables', 'factories'], true)) {
```
Array of keys is now used in two places, so we might want to extract it somewhere or improve this logic a bit.

The key observation here is that we want to apply normal replacement rules on all other keys than `aliases`, `invokables`, `factories`.

boesing replied:
Moved em into a constant.
Can change to a private static if we dont want to spread this to public.

michalbundyra commented on test/TestAsset/ConfigPostProcessor/FullServiceManagerConfiguration.php:
```diff
@@ -0,0 +1,42 @@
+            // NOTE: this is an invalid configuration, you have to provide service instances in this array.
+            // For testing purposes, we have to change this to a string we can use to verify that the value
+            // is not touched at all. Using instances would lead to
+            'Zend\Cache\Storage\Adapter\Redis' => 'Zend\Cache\Storage\Adapter\Redis',
```
As noted in the comment above the value here is wrong.
I believe non-string values should be left as-is (it should be verified).

Also: _I think_ as we are replacing key (alias?) we want to add alias (oldKey => newKey) at this point?

boesing replied:
Exactly, but I had to use as string to verify that its not gonna change.
Thus, in the `FullServiceManagerConfiguration.php.out`, I check if the value is still untouched (which is).

The unit test fails now as the special `services` handling is stripped away due to the only handling for `aliases`, `invokables` and `factories`.

boesing replied:
> Also: I think as we are replacing key (alias?) we want to add alias (oldKey => newKey) at this point?

Yeah, that would be a good one, if the alias does not exist.

boesing replied:
I've re-added the special `services` handling and adding an alias for a replaced service.

michalbundyra committed:
Removes redundant check

Signed-off-by: Michał Bundyra <contact@webimpress.com>

michalbundyra committed:
Reverts the previous logic

Signed-off-by: Michał Bundyra <contact@webimpress.com>

michalbundyra committed:
Reverts order of elements in array

Signed-off-by: Michał Bundyra <contact@webimpress.com>

boesing committed:
Fixed test asset

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

michalbundyra commented on src/ConfigPostProcessor.php:
```diff
@@ -251,9 +252,13 @@ private function replaceDependencyConfiguration(array $config)
+                continue;
+            }
+
+            $config[$key] = $this->__invoke($config[$key], [$key]);
```
we need to check here  - invoke only if `$config[$key]` is array, otherwise .... something else?

boesing committed:
Extended `ServiceManager` test

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
Added fix for non-array values inside `ServiceManager` configurations

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
Moved redundand `ServiceManager` keys to constant

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
Revert "Fixed test asset"

This reverts commit 810c2e4e

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing commented:
> @boesing Travis is failing, and it looks like delegator lists are getting **both** the translated Laminas class references **and** the original ZF ones. Can you review and either update test expectations (and, if you do, maybe indicate _why_) to include them, or fix the underlying problem (assuming you determine what it is).

Can you point me to the part in the configuration "out" file where this happens?
The delegators do not contain the old ZF reference.
Probably that was before the changes of @michalbundyra

boesing committed:
Added failing unit test for rewritten `services` key

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
Added special `ServiceManager::services` handling

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
Fixed codestyle and changed visibility of constant to private due to changing it to a private static property

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

michalbundyra commented on src/ConfigPostProcessor.php:
```diff
@@ -14,13 +14,17 @@
 class ConfigPostProcessor
 {
+    /** @var array<string> */
+    private static $SERVICE_MANAGER_KEYS_OF_INTEREST = ['aliases', 'invokables', 'factories', 'services'];
```
I prefer using constant here, we can mark it as internal, as it cannot be private (because of PHP 5.6).

Also, I would change it to:
```php
/**
 * @internal
 */
const SERVICE_MANAGER_KEYS_OF_INTEREST = [
    'aliases' => true,
    'factories' => true,
    'invokables' => true,
    'services' => true,
];
```

we can skip `@var` with type for constant I believe.

Using this format simplify the usage latter - we don't need to use `array_filp` and instead of `in_array` we can use `isset`.

boesing replied:
By making it a constant, the codestyle sniff is still complying with a warning. 
Not sure if thats still the case for `@internal` but thats why I've changed it to a private static property tho.

michalbundyra replied:
Check out my commit: https://github.com/boesing/laminas-zendframework-bridge/commit/3dc51591681f6da7e41a1b738ca0acbd6aa8308d

I've excluded one PSR-12 rule, as we still on PHP 5.6 here.

michalbundyra commented on src/ConfigPostProcessor.php:
```diff
@@ -75,7 +79,7 @@ function ($value, array $keys) {
             // service- and pluginmanager handling
             function ($value) {
-                $keysOfInterest = ['aliases', 'invokables', 'factories'];
+                $keysOfInterest = self::$SERVICE_MANAGER_KEYS_OF_INTEREST;
```
I believe we can skip assigning this variable now

michalbundyra commented on src/ConfigPostProcessor.php:
```diff
@@ -373,27 +390,28 @@ private function replaceDependencyFactories(array $config)
     {
-        if (empty($delegators)) {
-            return $delegators;
+        if (empty($config['services'])) {
```
I think we should play safe here, and add condition `|| ! is_array($config['services'])`

michalbundyra committed:
Use internal constant instead of private static property

Disabled PSR12.Properties.ConstantVisibility.NotFound CS rule, as we
still on PHP 5.6 in this library.

Signed-off-by: Michał Bundyra <contact@webimpress.com>

michalbundyra committed:
Adds condition to skip processing if the configuration is not an array

Signed-off-by: Michał Bundyra <contact@webimpress.com>

michalbundyra commented:
I've pushed two commits here but I can't see them above... :( github issue?
Please see: https://github.com/laminas/laminas-zendframework-bridge/compare/master...boesing:hotfix/missing-dependency-configuration-keys?expand=1

boesing committed:
Changed `ServiceManager` test to reflect a better real-world example

Contains valid values now

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
Added failing `ServiceManager::services` test when using nested array with details about zend-classes

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
Apply replacements on nested `ServiceManager::services` array configurations

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
Added invalid `ServiceManager::services` configuration

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
Do not handle invalid `ServiceManager::services` configuration

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
Added unit test for `services` instances

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
Added unit test for invalid configurations

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
Ensure that we only convert aliases if they're an array

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

michalbundyra commented on src/ConfigPostProcessor.php:
```diff
@@ -244,16 +250,24 @@ private function replaceExactValue($value)
     {
-        $aliases = isset($config['aliases']) ? $this->replaceDependencyAliases($config['aliases']) : [];
+        $aliases = isset($config['aliases']) && is_array($config['aliases']) ?
+            $this->replaceDependencyAliases($config['aliases']) : [];
```
formatting, please:
```php
$var = $condition
    ? $ifTrue
    : $ifFalse;
```

michalbundyra commented on test/ConfigPostProcessorTest.php:
```diff
@@ -15,6 +15,7 @@
 
 class ConfigPostProcessorTest extends TestCase
 {
+
```
it should not be allowed by CS, please revert

michalbundyra commented on test/ConfigPostProcessorTest.php:
```diff
@@ -38,12 +43,83 @@ public function configurations()
-        $processor              = new ConfigPostProcessor();
+        $config = require $configLocation;
+        $expected = require $expectedResultLocation;
+        $processor = new ConfigPostProcessor();
```
Please revert changes above

boesing committed:
Applied proper formatting

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
Added test for invalid `aliases` configuration and added a check for this

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

michalbundyra committed:
CS fixes

Signed-off-by: Michał Bundyra <contact@webimpress.com>

michalbundyra reviewed (approved):
LGTM 👍
